### PR TITLE
Added example of generating docs

### LINKED ISSUE #366: Implementation of OpenAPI for the api server
The benefits we want to achieve which this issue:
- generate documentation for the api
- generate api libraries (golang, javascript, python)

### PR DESCRIPTION
**Pull request by andriisoldatenko**
closes #366

### CODE REVIEW

File: api/Gopkg.toml [hunk 1]
```diff
@@ -46,3 +46,12 @@
 [[constraint]]
   name = "github.com/aws/aws-sdk-go"
   version = "1.10.24"
+
+[[constraint]]
+  name = "github.com/yvasiyarov/swagger"
+
+[[constraint]]
```
Comment (guusvw) [inline api/Gopkg.toml:53]: Why do you add this to the vendors?

File: api/Gopkg.toml [hunk 2]
```diff
@@ -46,3 +46,12 @@
...
+  name = "github.com/fvbock/endless"
+
+[[constraint]]
```
Comment (guusvw) [inline api/Gopkg.toml]: Why do you add this to the vendors?

File: api/hack/run-api.sh [hunk 3]
```diff
@@ -7,9 +7,11 @@ set -o pipefail
 
 cd ${GOPATH}/src/github.com/kubermatic/kubermatic/api
 ./_build/kubermatic-api \
-  -master-kubeconfig=../secrets/seed-clusters/dev.kubermatic.io/kubeconfig \
-  -datacenters=../secrets/seed-clusters/dev.kubermatic.io/datacenters.yaml \
-  -kubeconfig=../secrets/seed-clusters/dev.kubermatic.io/kubeconfig \
+  -master-kubeconfig=$GOPATH/src/github.com/kubermatic/secrets/seed-clusters/dev.kubermatic.io/kubeconfig \
+  -datacenters=$GOPATH/src/github.com/kubermatic/secrets/seed-clusters/dev.kubermatic.io/datacenters.yaml \
+  -kubeconfig=$GOPATH/src/github.com/kubermatic/secrets/seed-clusters/dev.kubermatic.io/kubeconfig \
+  -versions=../config/kubermatic/static/master/versions.yaml \
```
Comment (guusvw) [inline api/hack/run-api.sh]: Can you use an absolute path?

File: api/handler/handler.go [hunk 4]
```diff
@@ -26,3 +27,15 @@ func encodeJSON(c context.Context, w http.ResponseWriter, response interface{})
 	w.Header().Set("Content-Type", "application/json")
 	return json.NewEncoder(w).Encode(response)
 }
+
+// APIDescriptionHandler always return swagger index.json
+func APIDescriptionHandler(res http.ResponseWriter, _ *http.Request) {
+
+	f, e := ioutil.ReadFile("../api/handler/swagger/api/index.json")
```
Comment (guusvw) [inline api/handler/handler.go]: An parameter for the path to the file would be nice.

File: api/handler/handler.go [hunk 5]
```diff
@@ -26,3 +27,15 @@ func encodeJSON(c context.Context, w http.ResponseWriter, response interface{})
...
+	if e != nil {
+		res.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	res.Header().Set("Content-Type", "application/json")
+	json.NewEncoder(res).Encode(f)
```
Comment (guusvw) [inline api/handler/handler.go]: It's fine for me, but the `errcheck` don't like it, because you don't handle the error.

File: build_docs.sh [hunk 6]
```diff
@@ -0,0 +1,3 @@
+swagger -apiPackage="github.com/kubermatic/api" \
+    -mainApiFile="/Users/andrii/workspace/src/github.com/kubermatic/api/cmd/kubermatic-api/main.go" \
```
Comment (guusvw) [inline build_docs.sh]: Please use an enviroment variable like `$GOPATH`

File: docs/api-docs.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#### 1. Install & compile swagger
+```
+go get github.com/yvasiyarov/swagger
+go install
+```
+
+#### 2. Run swagger generator
+```
+swagger -apiPackage="github.com/kubermatic/api/cmd/kubermatic-api" \
+-mainApiFile="/Users/andrii/workspace/src/github.com/kubermatic/api/cmd/kubermatic-api/main.go" \
```
Comment (guusvw) [inline docs/api-docs.md]: Please use an enviroment variable like $GOPATH

File: api/cmd/kubermatic-api/main.go [hunk 8]
```diff
@@ -1,3 +1,11 @@
+// @APIVersion 1.4.0
+// @APITitle Kubermatic REST API
+// @APIDescription Kubermatic REST API
+// @BasePath http://loodse.com/
+// @Contact andrii.soldatenko@gmail.com
```
Comment (guusvw) [inline api/cmd/kubermatic-api/main.go]: You can use `info@loodse.com` as contact mail

File: api/cmd/kubermatic-api/main.go [hunk 9]
```diff
@@ -1,3 +1,11 @@
...
+// @TermsOfServiceUrl https://github.com/kubermatic/api
```
Comment (guusvw) [inline api/cmd/kubermatic-api/main.go]: You need to update the URL

**Review by guusvw — CHANGES_REQUESTED** [review]

File: api/handler/handler.go (hunk 4 above)
Comment (guusvw) [reply api/handler/handler.go]: That's not the "idiomatic go" naming convention, I would prefer `err`

File: api/handler/handler.go [hunk 10]
```diff
@@ -26,3 +27,15 @@ func encodeJSON(c context.Context, w http.ResponseWriter, response interface{})
 	w.Header().Set("Content-Type", "application/json")
 	return json.NewEncoder(w).Encode(response)
 }
+
+// APIDescriptionHandler always return swagger index.json
+func APIDescriptionHandler(res http.ResponseWriter, _ *http.Request) {
```
Comment (guusvw) [inline api/handler/handler.go]: That's not the "idiomatic go" naming convention, I would prefer `w`, like in the rest of the handler

File: api/handler/routing.go [hunk 11]
```diff
@@ -50,12 +50,12 @@ func NewRouting(
...
 }
 
-// Register registers all known endpoints in the given router.
+// Register declare router paths
 func (r Routing) Register(mux *mux.Router) {
 	mux.
 		Methods(http.MethodGet).
 		Path("/").
-		HandlerFunc(StatusOK)
+		HandlerFunc(APIDescriptionHandler)
```
Comment (guusvw) [inline api/handler/routing.go]: I'm not quite sure, but I think, we should'nt change the behaviour of the root handler.
@mrIncompetent do we use it for checks like an uptime check?

**Review by guusvw — CHANGES_REQUESTED** [review]

File: api/handler/routing.go (hunk 11 above)
Comment (mrIncompetent) [reply api/handler/routing.go]: Yes, please do not override this

File: api/Gopkg.toml (hunk 1 above)
Comment (andriisoldatenko) [reply api/Gopkg.toml:53]: @guusvw because this tool we use for generate `index.json`.

File: api/cmd/kubermatic-api/main.go (hunk 9 above)
Comment (andriisoldatenko) [reply api/cmd/kubermatic-api/main.go]: @guusvw do we have Terms Of Service Url?

File: api/handler/routing.go (hunk 11 above)
Comment (andriisoldatenko) [reply api/handler/routing.go]: what is the differences for `/healthz` or `/api/healthz`?
cc @mrIncompetent @scheeles @realfake

Comment (guusvw) [reply api/handler/routing.go]:
```
--- FAIL: TestUpRoute (0.00s)
        routing_test.go:99: Expected route to return code 200, got 500
```

if you don't have the generated docu, this test fails

Comment (andriisoldatenko) [reply api/handler/routing.go]: okey, I'll fix.

**Comment by guusvw** [comment]
Can not generate docu, get an error
```
~/g/s/g/k/k/api (feature/open-api-for-the-api-server ⚡↩☡-) ./hack/gen-api-docs.sh                                                                                                         12:43:00
2017/08/18 12:43:06 Start parsing
2017/08/18 12:43:08 Can not find definition of APIError model. Current package github.com/kubermatic/kubermatic/api/handler
```

File: api/handler/handler.go (hunk 4 above)
Comment (andriisoldatenko) [reply api/handler/handler.go]: @guusvw fixed.

**Comment by andriisoldatenko** [comment]
@guusvw try now to generate docs,
```
➜  api git:(feature/open-api-for-the-api-server) ✗ ./hack/gen-api-docs.sh
2017/08/18 15:08:36 Start parsing
2017/08/18 15:08:39 Finish parsing
2017/08/18 15:08:39 Wrote api/index.json
2017/08/18 15:08:39 Swagger UI files generated
```

File: api/Gopkg.toml (hunk 2 above)
Comment (andriisoldatenko) [reply api/Gopkg.toml]: removed

File: docs/api-docs.md (hunk 7 above)
Comment (andriisoldatenko) [reply docs/api-docs.md]: fixed

File: api/hack/run-api.sh (hunk 3 above)
Comment (andriisoldatenko) [reply api/hack/run-api.sh]: fixed

File: api/handler/handler.go [hunk 12]
```diff
@@ -36,13 +36,17 @@ func encodeJSON(c context.Context, w http.ResponseWriter, response interface{})
...
+func APIDescriptionHandler(w http.ResponseWriter, r *http.Request) {
 
 	f, err := ioutil.ReadFile("../api/handler/swagger/api/index.json")
 	if err != nil {
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 	w.Header().Set("Content-Type", "application/json")
-	w.Write(f)
+	_, err1 := w.Write(f)
```
Comment (guusvw) [inline api/handler/handler.go]: `_, err = w.Write(f)` - works fine too ;)

File: api/cmd/kubermatic-api/main.go (hunk 9 above)
Comment (guusvw) [reply api/cmd/kubermatic-api/main.go]: I don't think so. So, you can remove the TOS tag

**Review by guusvw — CHANGES_REQUESTED** [review]

**Review by guusvw — APPROVED** [review]

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**